lit/160_parameters.md
# Parameters Encoding/Decoding

## Overview

### Motivation

The machine generated part of the brainkey (the bk-passphrase) is
prefixed with the kdf parameters in an encoded form. These
parameters are used to derive the wallet seed.

The parameters used to derive a wallet seed are critical. We
would like to avoid usability issues such as with derivation
paths of deterministic wallets, which have implicit parameters
that [may differ from one wallet implementation to the
next][href_derivation_paths].

Since the user must remember these parameters, we piggy-back on
the other thing they must remember anyway, namely the
bk-passphrase. This way, they have no extra conceptual burden, no
separate thing they need to keep track of. The user documentation
should also not state this implementation detail to users, so as
to minimize cognitive load that might scare away marginal users.

As the parameters must be memorized, a compact encoding is
critical. For example, we reserve only two bits for a version
number, to allow for iteration of the implementation. If further
iterations are needed in the future, then user facing changes
will also be needed.

[href_derivation_paths]: https://walletsrecovery.org/


### Data layout

```bob
        kdf          share
    .---------. .-------------.
    +         + +             +
0 1 2 3 4 5 6 7 8 9 A B C D E F
+-+ +---+ +---+ +-------+ +---+
+-+ +---+ +---+ +-------+ +---+
 v    m     t       x       t'
```

|      | identifier | bits | description |
|------|------------|------|-------------|
| `v`  | `version`  | `2`  |             |
| `m`  | `kdf_m`    | `3`  |             |
| `t`  | `kdf_t`    | `3`  |             |
| `x`  | `sss_x`    | `5`  |             |
| `t'` | `sss_t`    | `3`  |             |


### Dataclass: `Parameters`

Throughout SBK, the decoded parameters are passed using an instance
of `Parameters`.

```python
# def: type_parameters
class Parameters(NamedTuple):

    version   : int

    kdf_p: ct.Parallelism
    kdf_m: ct.MebiBytes
    kdf_t: ct.Iterations

    sss_x: int
    sss_t: int
    sss_n: int
```

The `sss_*` parameters may not always be available:
- `sss_x` may be `-1` when params were decoded from a salt
- `sss_n` are usually `-1` except when generating shares.

In addition we use a subset of the parameters for any context that is
strictly concerned with key derivation and has nothing to do with
shamir shares.

```python
# def: type_kdf_params
class KDFParams(NamedTuple):
    kdf_p: ct.Parallelism
    kdf_m: ct.MebiBytes
    kdf_t: ct.Iterations
```


### Module: `sbk.parameters`

```python
# file: src/sbk/parameters.py
# include: common.boilerplate
# dep: common.imports, constant*, type*, impl*
```

```bash
# run: bash scripts/lint.sh src/sbk/parameters.py
# exit: 0
```


### Constants for KDF

```python
# def: constants
SBK_VERSION_V0 = 0

# constrained by f_threshold (3bits)
MIN_THRESHOLD = 2
MAX_THRESHOLD = 10

KDF_PARALLELISM       = ct.Parallelism(128)   # hardcoded
DEFAULT_KDF_T_TARGET  = ct.Seconds(90)
DEFAULT_KDF_M_PERCENT = 100

DEFAULT_SSS_T = 3
DEFAULT_SSS_N = 5

V0_KDF_M_BASE = 1.5
V0_KDF_T_BASE = 4.0

V0_KDF_M_UNIT = 512     # megabytes
V0_KDF_T_UNIT = 1       # iterations
```


### Constants for Secrets

The length of a share consists of the ``parameters + brainkey +
salt_hash``. This gives us a total of 24 bytes/words. The values were
chosen with the following priority of constraints:

1. Due to encoding constraints, the header length for a share is fixed
   at 2 bytes.
2. The main constraint on the brainkey is a minimum entropy to protect
   against a compromised salt. This must be balanced against the
   maximum number of words a human can be expected to memorize.
3. With the previous two constraints on the header and brainkey, any
   remaining constraints must be satisfied by the salt. The main
   constraint here is a minimum level of total entropy.

With an entropy of at least 5 words/bytes = 40 bits, the brainkey
is expensive but perhaps not infeasible to brute force. This low
value is only justified as the attack to defend against is the
narrow case of a compromised salt. The wallet owner is intended
ot be the only person who knows the salt. Should they reveal the
salt by accident, they will still have time to transfer their
funds to a safe wallet.

The resulting total entropy is on the order of ``11 + 5 = 16byte =
128bit``, depending on how well the user chose their salt. The
header is not completely random, so we cannot count them as part
of the entropy.

```python
# def: constants_lens
BRANKEY_HEADER_LEN = 1
SHARE_HEADER_LEN   = 2

DEFAULT_RAW_SALT_LEN = 11
DEFAULT_BRAINKEY_LEN = 5
```


lit/180_parameters.md
# Parameters Encoding/Decoding

## Overview

### Data layout

Each Salt and Share is prefixed with the encoded parameters. These
parameters are used to derive the wallet seed. For the salt, the
header is 2 bytes, for each share it is 3 bytes.


```bob
     share            "kdf_params"
.-------------. .---------------------.
1             1 0                             0
7 6 5 4 3 2 1 0 F E D C B A 9 8 7 6 5 4 3 2 1 0
+-------+ +---+ +---------+ +---------+ . +---+
+-------+ +---+ +---------+ +---------+ ' +---+
   :        :       :           :       :   :
   '        :       '           '       :   :
  "sss_x"   :     "kdf_t"    "kdf_m"    :   :
  "(5bits)" :     "(6bits)"  "(6bits)"  :   :
            '                           :   :
         "sss_t"              .~~~~~~~~~'   '
         "(3bits)"         "paranoid"     "version"
                           "(1bit)"       "(3bits)"
```


### Data Packing

A short interlude on how packing/unpacking binary data works. We want
to encode data in three bytes and our choice of endianness should
reflect the documented data layout.

```python
# exec
from struct import unpack

result, = unpack("!L", b"\x00\xFE\xDC\xBA")
assert (0xFE << 16) | (0xDC << 8) | (0xBA << 0) == result
```


### Module: `sbk.parameters`

```python
# file: src/sbk/parameters.py
# include: common.boilerplate
# dep: common.imports, constants, types, impl
```

```bash
# run: bash scripts/lint.sh src/sbk/parameters.py
# exit: 0
```


### Constants

```python
# def: parameters.constants
SBK_VERSION_V0 = 0

# constrained by f_threshold (3bits)
MIN_THRESHOLD = 2
MAX_THRESHOLD = 10

DEFAULT_RAW_SALT_LEN  = 7
PARANOID_RAW_SALT_LEN = 13

DEFAULT_BRAINKEY_LEN  = 6
PARANOID_BRAINKEY_LEN = 8

PARAM_DATA_LEN = 2
SHARE_DATA_LEN = 3

KDF_PARALLELISM = 128   # hardcoded
DEFAULT_KDF_T_TARGET = 90   # seconds

DEFAULT_SSS_T  = 3
DEFAULT_SSS_N = 5
```


### Dataclass: `Parameters`

```python
# def: types
class Parameters(NamedTuple):

    version   : int
    paranoid  : bool

    kdf_p: ct.Parallelism
    kdf_m: ct.MebiBytes
    kdf_t: ct.Iterations

    sss_x: int
    sss_t: int
    sss_n: int
```

The `sss_*` parameters may not always be available:
- `sss_x` may be `-1` when params were decoded from a salt
- `sss_n` are usually `-1` except when generating shares.


CHANGELOG.md
# Changelog for https://gitlab.com/mbarkhau/sbk

## v201912.1001-alpha

 - Initial release


src/sbk/parameters.py
#################################################
#     This is a generated file, do not edit     #
#################################################

# This file is part of the sbk project
# https://github.com/mbarkhau/sbk
#
# Copyright (c) 2019-2022 Manuel Barkhau (mbarkhau@gmail.com) - MIT License
# SPDX-License-Identifier: MIT
import os
import re
import sys
import json
import math
import time
import base64
import struct
import typing as typ
import hashlib
import logging
import pathlib as pl
import functools as ft
import itertools as it
import threading
import subprocess as sp
from typing import Any
from typing import Set
from typing import Dict
from typing import List
from typing import Type
from typing import Tuple
from typing import Union
from typing import Generic
from typing import NewType
from typing import TypeVar
from typing import Callable
from typing import Iterable
from typing import Iterator
from typing import Optional
from typing import Protocol
from typing import Sequence
from typing import Generator
from typing import NamedTuple

# from collections.abc import Generator, Iterator, Counter

# from typing import TypeAlias
TypeAlias = Any

import sbk.common_types as ct

logger = logging.getLogger(__name__)
SBK_VERSION_V0 = 0

# constrained by f_threshold (3bits)
MIN_THRESHOLD = 2
MAX_THRESHOLD = 10

KDF_PARALLELISM = ct.Parallelism(128)  # hardcoded
DEFAULT_KDF_T_TARGET = ct.Seconds(90)
DEFAULT_KDF_M_PERCENT = 100

DEFAULT_SSS_T = 3
DEFAULT_SSS_N = 5

V0_KDF_M_BASE = 1.5
V0_KDF_T_BASE = 4.0

V0_KDF_M_UNIT = 512  # megabytes
V0_KDF_T_UNIT = 1  # iterations
BRANKEY_HEADER_LEN = 1
SHARE_HEADER_LEN = 2

DEFAULT_RAW_SALT_LEN = 11
DEFAULT_BRAINKEY_LEN = 5
if "SBK_DEBUG_RAW_SALT_LEN" in os.environ:
    DEFAULT_RAW_SALT_LEN = int(os.environ["SBK_DEBUG_RAW_SALT_LEN"])

if "SBK_DEBUG_BRAINKEY_LEN" in os.environ:
    DEFAULT_BRAINKEY_LEN = int(os.environ["SBK_DEBUG_BRAINKEY_LEN"])

MIN_ENTROPY = int(os.getenv("SBK_MIN_ENTROPY", "16"))
MAX_ENTROPY_WAIT = int(os.getenv("SBK_MAX_ENTROPY_WAIT", "10"))

DEFAULT_KDF_T_TARGET = int(os.getenv("SBK_KDF_T_TARGET") or DEFAULT_KDF_T_TARGET)
DEFAULT_KDF_M_PERCENT = int(os.getenv("SBK_KDF_M_PERCENT") or DEFAULT_KDF_M_PERCENT)

DEFAULT_SSS_T = int(os.getenv("SBK_THRESHOLD") or DEFAULT_SSS_T)
DEFAULT_SSS_N = int(os.getenv("SBK_NUM_SHARES") or DEFAULT_SSS_N)


class Parameters(NamedTuple):

    version: int

    kdf_p: ct.Parallelism
    kdf_m: ct.MebiBytes
    kdf_t: ct.Iterations

    sss_x: int
    sss_t: int
    sss_n: int


class KDFParams(NamedTuple):
    kdf_p: ct.Parallelism
    kdf_m: ct.MebiBytes
    kdf_t: ct.Iterations


def param_coeffs(b: float) -> Tuple[int, int]:
    assert b > 1
    s = 1 / (b - 1)
    o = 1 - s

    v0 = b ** 0 * s + o
    v1 = b ** 1 * s + o
    assert v0 == 1
    assert 1.5 < v1 < 2.5
    return (s, o)


from math import log


def param_exp(n: int, b: float) -> int:
    s, o = param_coeffs(b)
    v = round(b ** n * s + o)
    return v


def param_log(v: int, b: float) -> int:
    s, o = param_coeffs(b)
    n = log((v - o) / s) / log(b)
    return min(max(round(n), 0), 2 ** 63)


def init_kdf_params(kdf_m: ct.MebiBytes, kdf_t: ct.Iterations) -> KDFParams:
    kdf_m_enc = param_log(kdf_m / V0_KDF_M_UNIT, V0_KDF_M_BASE)
    kdf_t_enc = param_log(kdf_t / V0_KDF_T_UNIT, V0_KDF_T_BASE)

    kdf_m = param_exp(kdf_m_enc, V0_KDF_M_BASE) * V0_KDF_M_UNIT
    kdf_t = param_exp(kdf_t_enc, V0_KDF_T_BASE) * V0_KDF_T_UNIT
    return KDFParams(KDF_PARALLELISM, kdf_m, kdf_t)


def init_parameters(
    kdf_m: ct.MebiBytes,
    kdf_t: ct.Iterations,
    sss_x: int,
    sss_t: int = DEFAULT_SSS_T,
    sss_n: int = -1,
) -> Parameters:
    kdf_params = init_kdf_params(kdf_m, kdf_t)
    if not MIN_THRESHOLD <= sss_t <= MAX_THRESHOLD:
        raise ValueError(f"Invalid threshold: {sss_t}")
    elif kdf_params.kdf_m % V0_KDF_M_UNIT != 0:
        errmsg = f"Invalid kdf_m: {kdf_params.kdf_m} % {V0_KDF_M_UNIT} != 0"
        raise ValueError(errmsg)
    else:
        return Parameters(
            version=SBK_VERSION_V0,
            kdf_p=kdf_params.kdf_p,
            kdf_m=kdf_params.kdf_m,
            kdf_t=kdf_params.kdf_t,
            sss_x=sss_x,
            sss_t=sss_t,
            sss_n=sss_n,
        )


def params2bytes(params: Parameters) -> bytes:
    assert params.version == 0

    kdf_m_enc = param_log(params.kdf_m / V0_KDF_M_UNIT, V0_KDF_M_BASE)
    kdf_t_enc = param_log(params.kdf_t / V0_KDF_T_UNIT, V0_KDF_T_BASE)

    assert params.version & 0b0011 == params.version
    assert kdf_m_enc & 0b0111 == kdf_m_enc
    assert kdf_t_enc & 0b0111 == kdf_t_enc

    if params.sss_x > 0:
        sss_x_enc = params.sss_x - 1
    else:
        sss_x_enc = 0
    sss_t_enc = params.sss_t - 2

    assert sss_x_enc & 0b0001_1111 == sss_x_enc
    assert sss_t_enc & 0b0000_0111 == sss_t_enc

    field0 = 0 | params.version << 0x00 | kdf_m_enc << 0x02 | kdf_t_enc << 0x05
    field1 = 0 | sss_x_enc << 0x00 | sss_t_enc << 0x05
    return struct.pack("<BB", field0, field1)


def bytes2params(data: bytes) -> Parameters:
    is_share_data = len(data) == 2
    if not is_share_data:
        data = data + b"\x00"  # append dummy sss_t and sss_x

    assert len(data) == 2, len(data)
    field0, field1 = struct.unpack("<BB", data)

    version = (field0 >> 0x00) & 0b0011
    kdf_m_enc = (field0 >> 0x02) & 0b0111
    kdf_t_enc = (field0 >> 0x05) & 0b0111

    sss_x_enc = (field1 >> 0x00) & 0b0001_1111
    sss_t_enc = (field1 >> 0x05) & 0b0000_0111

    assert version == SBK_VERSION_V0, f"Invalid version: {version}"

    kdf_m = param_exp(kdf_m_enc, V0_KDF_M_BASE) * V0_KDF_M_UNIT
    kdf_t = param_exp(kdf_t_enc, V0_KDF_T_BASE) * V0_KDF_T_UNIT
    if is_share_data:
        sss_x = sss_x_enc + 1
        sss_t = sss_t_enc + 2
    else:
        sss_x = -1
        sss_t = 2

    sss_n = sss_t
    return init_parameters(kdf_m, kdf_t, sss_x, sss_t, sss_n)


def validated_param_data(params: Parameters) -> bytes:
    # validate encoding round trip before we use params
    params_data = params2bytes(params)
    decoded_params = bytes2params(params_data)

    checks = {
        "threshold": params.sss_t == decoded_params.sss_t,
        "version": params.version == decoded_params.version,
        "kdf_p": params.kdf_p == decoded_params.kdf_p,
        "kdf_m": params.kdf_m == decoded_params.kdf_m,
        "kdf_t": params.kdf_t == decoded_params.kdf_t,
    }
    bad_checks = [name for name, is_ok in checks.items() if not is_ok]
    if any(bad_checks):
        raise ValueError(bad_checks)
    else:
        return params_data


class SecretLens(NamedTuple):
    raw_brainkey: int
    raw_share: int
    brainkey: int
    salt: int
    master_key: int
    share: int


def raw_secret_lens() -> SecretLens:
    raw_brainkey = DEFAULT_BRAINKEY_LEN
    salt = DEFAULT_RAW_SALT_LEN
    raw_share = raw_brainkey + salt

    brainkey = BRANKEY_HEADER_LEN + raw_brainkey
    master_key = raw_brainkey + salt
    share = SHARE_HEADER_LEN + raw_share
    return SecretLens(**locals())
